src/devices/sound/qsoundhle.cpp

```diff
@@ -180,70 +212,429 @@ WRITE8_MEMBER(qsound_hle_device::qsound_w)
 
 READ8_MEMBER(qsound_hle_device::qsound_r)
 {
-	/* Port ready bit (0x80 if ready) */
-	return 0x80;
+	// ready bit (0x00 = busy, 0x80 == ready)
+	m_stream->update();
+	return m_ready_flag;
 }
 
 
 void qsound_hle_device::write_data(u8 address, u16 data)
 {
-	int ch = 0, reg;
+	uint16_t *destination = register_map[address];
+	if(destination)
+		*destination = data;
+	m_ready_flag = 0;
+}
 
-	if (address < 0x80)
+void qsound_hle_device::init_register_map()
+{
+	// unused registers
+	for(int i=0;i<256;i++)
+		register_map[i] = NULL;
```

Comment:
Please use `nullptr` as a null pointer constant.  Also, probably a case for `std::fill`